backend/algorithms/autoplanning.py

```diff
@@ -0,0 +1,114 @@
+            boolean_indexes.append(b)
+        # if the course is in term 'index', only allow 0 to m UOC to exist in that term. 
+        model.AddReservoirConstraintWithActive(
+            variables,
+            list(map_var_to_course(courses, var).uoc for var in variables), # this fills the resovoir by UoC units if active
+            boolean_indexes, # a course is only active if in term 'index'
+            0,
+            m # uoc_max
+        )
+
+    # enforce prereqs
+    for course in courses:
+        # this is the responsibility of the condition class to generate this.
+        course.condition.condition_to_model(
+            model,
+            user,
+            list((variable, map_var_to_course(courses, variable)) for variable in variables),
+            map_course_to_var(course, variables)
+        )
+    solver = cp_model.CpSolver()
+    status: int = solver.Solve(model)
+    if status == 3:
+        raise Exception('your courses are impossible to put in these terms!')
+    else:
+        return {v.Name(): convert_to_term_year(solver.Value(v), start) for v in variables}
+
+
+if __name__ == '__main__':
+    pprint(autoplan(
+        [
+            Course("MATH1141", CONDITIONS["MATH1141"], 65, 6, {2020: [1, 3]}),
+            Course("MATH1081", CONDITIONS["MATH1081"], 65, 6, {2020: [1, 2, 3]}),
+            Course("COMP1511", CONDITIONS["COMP1511"], 65, 6, {2020: [1, 2, 3]}),
+            Course("COMP2521", CONDITIONS["COMP2521"], 65, 6, {2020: [2, 3]}),
+            Course("COMP2041", CONDITIONS["COMP2041"], 65, 6, {2020: [2]}),
+            Course("COMP1531", CONDITIONS["COMP1531"], 65, 6, {2020: [1, 3]}),
+            Course("COMP1521", CONDITIONS["COMP1521"], 65, 6, {2020: [1, 2]}),
+            Course("ENGG2600", CONDITIONS["ENGG2600"], 65, 2, {2021: [1, 2, 3]}),
+            Course("ENGG2600", CONDITIONS["ENGG2600"], 65, 2, {2021: [1, 2, 3]}),
+            Course("ENGG2600", CONDITIONS["ENGG2600"], 65, 2, {2021: [1, 2, 3]}),
+            Course("COMP2511", CONDITIONS["COMP2511"], 65, 6, {2020: [2, 3], 2021: [2, 3]}),
+            Course("MATH1241", CONDITIONS["MATH1141"], 65, 6, {2020: [2, 3]}),
+            Course("MATH3411", CONDITIONS["MATH3411"], 65, 6, {2020: [3]}),
+            Course("COMP3411", CONDITIONS["COMP3411"], 65, 6, {2021: [0]}),
+            Course("COMP6841", CONDITIONS["COMP6841"], 65, 6, {2021: [1]}),
+            Course("COMP3231", CONDITIONS["COMP3231"], 65, 6, {2021: [1]}),
+            Course("COMP3141", CONDITIONS["COMP3141"], 65, 6, {2021: [2]}),
+            Course("COMP3121", CONDITIONS["COMP3121"], 65, 6, {2021: [2, 3]}),
+            Course("COMP3131", CONDITIONS["COMP3131"], 65, 6, {2022: [1]}),
+            Course("COMP4141", CONDITIONS["COMP4141"], 65, 6, {2022: [1]}),
+            Course("COMP3311", CONDITIONS["COMP3311"], 65, 6, {2022: [1, 2, 3]}),
+            Course("ARTS1360", CONDITIONS["ARTS1360"], 65, 6, {2022: [2]}),
+        ],
```

Comment:
I understand we trust the tool to work but can we provide a more comprehensive list of offerings? This is the one I used to test btw, sorry I have trust issues xD 
Everything is working fine except ENGG2600.

```
Course("MATH1141", CONDITIONS["MATH1141"], 65, 6, {2020: [1, 3], 2021: [1, 3], 2022: [1, 3]}),
Course("MATH1081", CONDITIONS["MATH1081"], 65, 6, {2020: [1, 2, 3], 2021: [1, 2, 3], 2022: [1, 2, 3]}),
Course("COMP1511", CONDITIONS["COMP1511"], 65, 6, {2020: [1, 2, 3], 2021: [1, 2, 3], 2022: [1, 2, 3]}),
Course("COMP2521", CONDITIONS["COMP2521"], 65, 6, {2020: [2, 3], 2021: [2, 3], 2022: [2, 3]}),
Course("COMP2041", CONDITIONS["COMP2041"], 65, 6, {2020: [2], 2021: [2], 2022: [2]}),
Course("COMP1531", CONDITIONS["COMP1531"], 65, 6, {2020: [1, 3], 2021: [1, 3], 2022: [1, 3]}),
Course("COMP1521", CONDITIONS["COMP1521"], 65, 6, {2020: [1, 2], 2021: [1, 2], 2022: [1, 2]}),
Course("ENGG2600", CONDITIONS["ENGG2600"], 65, 2, {2020: [1, 2, 3], 2021: [1, 2, 3], 2022: [1, 2, 3]}),
Course("ENGG2600", CONDITIONS["ENGG2600"], 65, 2, {2020: [1, 2, 3], 2021: [1, 2, 3], 2022: [1, 2, 3]}),
Course("ENGG2600", CONDITIONS["ENGG2600"], 65, 2, {2020: [1, 2, 3], 2021: [1, 2, 3], 2022: [1, 2, 3]}),
Course("COMP2511", CONDITIONS["COMP2511"], 65, 6, {2020: [2, 3], 2021: [2, 3], 2022: [2, 3]}),
Course("MATH1241", CONDITIONS["MATH1141"], 65, 6, {2020: [2, 3], 2021: [2, 3], 2022: [2, 3]}),
Course("MATH3411", CONDITIONS["MATH3411"], 65, 6, {2020: [3], 2021: [3], 2022: [3]}),
Course("COMP3411", CONDITIONS["COMP3411"], 65, 6, {2020: [3], 2021: [0], 2022: [0]}),
Course("COMP6841", CONDITIONS["COMP6841"], 65, 6, {2020: [1], 2021: [1], 2022: [1]}),
Course("COMP3231", CONDITIONS["COMP3231"], 65, 6, {2020: [1], 2021: [1], 2022: [1]}),
Course("COMP3141", CONDITIONS["COMP3141"], 65, 6, {2020: [2], 2021: [2], 2022: [2]}),
Course("COMP3121", CONDITIONS["COMP3121"], 65, 6, {2020: [2, 3], 2021: [2, 3], 2022: [2, 3]}),
Course("COMP3131", CONDITIONS["COMP3131"], 65, 6, {2020: [1], 2021: [1], 2022: [1]}),
Course("COMP4141", CONDITIONS["COMP4141"], 65, 6, {2020: [1], 2021: [1], 2022: [1]}),
Course("COMP3311", CONDITIONS["COMP3311"], 65, 6, {2020: [1, 2, 3], 2021: [1, 2, 3], 2022: [1, 2, 3]}),
Course("ARTS1360", CONDITIONS["ARTS1360"], 65, 6, {2020: [2], 2021: [2], 2022: [2]}),
```

Reply:
yup oops the bug was with printing, not the model